### DISCUSSION TITLE
Denoising SDXL iteration images for coherent image previews that a user could understand #7001

### DISCUSSION THREAD

**Discussion by MonkeeMan1** [huggingface/diffusers#7135] [Q&A]
Hello,

I'm currently trying to create image previews with SDXL. This works! However, the image output are very noisy. A very long time ago I found a solution to this for sd1.5 but unfortunately it has been lost to time. 

How would I go about denoising these images so they are a little more coherent to a human viewer? I know the first couple of iterations are always going to be very noisy, but eventually it should be possible to convert this noise into a blurry image that a human could understand. 

```python
import time
from diffusers import StableDiffusionXLPipeline
import torch

pipe = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"

def callback(pipe, step_index, timestep, callback_kwargs):
    latents = callback_kwargs.get("latents")

    start_time = time.time()
    with torch.no_grad():
        pipe.upcast_vae()
        latents = latents.to(
            next(iter(pipe.vae.post_quant_conv.parameters())).dtype)
        images = pipe.vae.decode(
            latents / pipe.vae.config.scaling_factor, return_dict=False)[0]
        images = pipe.image_processor.postprocess(images, output_type='pil')

        images[0].save(f"./imgs/{step_index}.png")

    end_time = time.time()

    print(f"Time taken to generate image: {end_time - start_time} seconds")

    return callback_kwargs

pipe(prompt=prompt, callback_on_step_end=callback)
```

The desired output can be seen below, turning the noisey image into a somewhat comprehensible image that a human can understand is a dog. (Both of these images were after 1 iteration)
![image](https://github.com/huggingface/diffusers/assets/98274841/7b8fb4dc-55d7-4b4e-bf1f-ed3fe9ef4e2b)

**Comment by asomoza** [comment]
I actually was thinking that this wasn't possible since the noisy images you're seeing is what's happening in real time, a simple blur doesn't get the result you want.

But I found something that could work for what you want, based in my answer from here: https://github.com/huggingface/diffusers/discussions/6991

I get this result (without taesd):

![preview_normal_1x](https://github.com/huggingface/diffusers/assets/5442875/d7df9283-eff7-439b-8f86-4d2e12e0c6d4)

There's a denoising technique that works with those images, the result is like this:

![preview_denoised_1x](https://github.com/huggingface/diffusers/assets/5442875/8ea67e44-f626-4a02-a462-461e6e6df305)

And it seems you don't care that much about the quality, so you can scale it up:

2X
![preview_denoised_2x](https://github.com/huggingface/diffusers/assets/5442875/333608da-9536-44a8-ba1b-bb8716429318)

4X
![preview_denoised_4x](https://github.com/huggingface/diffusers/assets/5442875/4001689f-4050-494f-88cd-a4df47c1beab)

This doesn't work with taesd that well, the noise is not the same, and also it's really slow (with taesd). 

If this works for you I can clean the code and paste it here, probably need to lower the strength towards the end to get a less blurry image as well.

**Comment by sayakpaul** [comment]
Cc: @yiyixuxu @stevhliu another technique from @asomoza <3
